feat(query) : Binary Join Many to Many - "and", "or" and "unless"

TanviBhavsar opened the pull request:
Support for Binary Join Many to Many with "and", "or" and "unless" operators

vishramachandran commented:
@TanviBhavsar after looking at the PR a bit more, I think I understand your motivation of colocating the two. Let me think a bit more.

vishramachandran commented on query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:
```diff
@@ -0,0 +1,142 @@
+                                cardinality: Cardinality,
+                                on: Seq[String],
+                                ignoring: Seq[String]) extends NonLeafExecPlan {
+  require(cardinality == Cardinality.ManyToMany, "set operations must only use many-to-many matching")
```
IS `cardinality` even needed now ? Can be removed.

TanviBhavsar replied:
Prometheus has this validation that set operations must only use many-to-many matching

TanviBhavsar replied:
Removed cardinality. Validation is done in vectorMatch validate

vishramachandran commented on query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:
```diff
@@ -0,0 +1,142 @@
+      // Add range vectors from lhs which are present in lhs and rhs both
+      // Result should also have range vectors for which rhs does not have any keys
+      if (rhsKeysSet.contains(jk) || rhsKeysSet.isEmpty) {
+        result += IteratorBackedRangeVector(lhs.key, lhs.rows)
```
No need to create new RV. Can simply used `lhs`. Same comment on a few other places.

TanviBhavsar replied:
I missed that. Made the change

vishramachandran commented on query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:
```diff
@@ -0,0 +1,142 @@
+      rhsKeysSet += jk
+    }
+    // Add range vectors which are not present in rhs
+    lhsRvs.foreach { lhs =>
```
consider using map/filter instead of foreach.

TanviBhavsar commented:
> @TanviBhavsar Can we simply use another ExecPlan for Set Operators ?
> They are completely different.
> 
> 1. We are trying to fit in Set operators as ManyToMany which is a bit kludgy
> 2. I am not seeing any logic that needs to be shared/reused between the two - the flow is fully disjoint.
> 
> I dont see why these need to be co-located.
> 
> Let me know your thoughts, and lets discuss in person if needed.

Have added Exec plan for Set Operator

TanviBhavsar commented:
> Thanks for the changes.
> Looks great. Some minor changes requested. Then we will be ready to go.
@vishramachandran  Addressed your review comments

TanviBhavsar commented on query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala:
```diff
@@ -42,8 +41,7 @@ final case class BinaryJoinExec(id: String,
                                 ignoring: Seq[String],
                                 include: Seq[String]) extends NonLeafExecPlan {
-  require(cardinality != Cardinality.ManyToMany || binaryOp.isInstanceOf[ComparisonOperator],
-    "Many To Many cardinality is supported only for comparison binary operators")
```
Modified the check as discussed
